src/resolve/resolved-dns-transaction.c

```diff
@@ -2055,40 +2055,40 @@ int dns_transaction_go(DnsTransaction *t) {
                 if (IN_SET(r, -EMSGSIZE, -EAGAIN, -EPERM))
                         r = dns_transaction_emit_tcp(t);
         }
-        if (r == -ELOOP) {
-                if (t->scope->protocol != DNS_PROTOCOL_DNS)
-                        return r;
+        if (r < 0) {
+                if (r == -ELOOP) {
+                        if (t->scope->protocol != DNS_PROTOCOL_DNS)
+                                return r;
 
-                /* One of our own stub listeners */
-                log_debug_errno(r, "Detected that specified DNS server is our own extra listener, switching DNS servers.");
+                        /* One of our own stub listeners */
+                        log_debug_errno(r, "Detected that specified DNS server is our own extra listener, switching DNS servers.");
 
-                dns_scope_next_dns_server(t->scope, t->server);
+                        dns_scope_next_dns_server(t->scope, t->server);
 
-                if (dns_scope_get_dns_server(t->scope) == t->server) {
-                        log_debug_errno(r, "Still pointing to extra listener after switching DNS servers, refusing operation.");
-                        dns_transaction_complete(t, DNS_TRANSACTION_STUB_LOOP);
+                        if (dns_scope_get_dns_server(t->scope) == t->server) {
+                                log_debug_errno(r, "Still pointing to extra listener after switching DNS servers, refusing operation.");
+                                dns_transaction_complete(t, DNS_TRANSACTION_STUB_LOOP);
+                                return 0;
+                        }
+
+                        return dns_transaction_go(t);
+                }
+                if (r == -ESRCH) {
+                        /* No servers to send this to? */
+                        dns_transaction_complete(t, DNS_TRANSACTION_NO_SERVERS);
+                        return 0;
+                }
+                if (r == -EOPNOTSUPP) {
+                        /* Tried to ask for DNSSEC RRs, on a server that doesn't do DNSSEC  */
+                        dns_transaction_complete(t, DNS_TRANSACTION_RR_TYPE_UNSUPPORTED);
+                        return 0;
+                }
+                if (t->scope->protocol == DNS_PROTOCOL_LLMNR && ERRNO_IS_DISCONNECT(r)) {
```

Comment:
Hmm... I'd like to keep the previous code, but only add `r < 0` before `ERRNO_IS_DISCONNECT(r)`:
```c
if (t->scope->protocol == DNS_PROTOCOL_LLMNR && r < 0 && ERRNO_IS_DISCONNECT(r))
```